README.md
# docker-logstash-kubernetes

Logstash container for pulling docker logs with kubernetes metadata support.
Additionally logs are pulled from systemd journal too.

Logstash tails docker logs and extracts `pod`, `container_name`, `namespace`,
etc. The way this works is very simple. Logstash looks at an event field which
contains full path to kubelet created symlinks to docker container logs, and
extracts useful information from a symlink name. No access to Kubernetes API
is required.

Other outputs can be added in the future.

## Requirements

You need to have kubelet process running on the host. Normally kubelet creates
symlinks to container logs from `/var/log/containers/` to
`/var/lib/docker/containers/`. So for that you need to make sure that logstash
has access to both directories.

For logstash to be able to pull logs from journal, you need to make sure that
logstash can read `/var/log/journal`.

Also, logstash writes `sincedb` file to its home directory, which by default is
`/var/lib/logstash`. If you don't want logstash to start reading docker or
journal logs from the beginning after a restart, make sure you mount
`/var/lib/logstash` somewhere on the host.

## Configuration

As usual, configuration is passed through environment variables.

- `LS_HEAP_SIZE` - Logstash JVM heap size. Defaults to `500m`.
- `LS_LOG_LEVEL` - Logstash log level. Default: `error`.
- `LS_NODE_NAME` - Logstash node name reported to ES if `LS_MONITORING_ENABLE=true`. Default: `$HOSTNAME`
- `LS_PIPELINE_BATCH_SIZE` - Size of batches the pipeline is to work in. Default: `125`
- `LS_MONITORING_ENABLE` - Whether to enable Logstash xpack monitoring. Default: `false`
- `LS_HTTP_HOST` - Listen address. default: `0.0.0.0`.
- `LS_JAVA_OPTS` - JVM Options. Default: `-Djava.io.tmpdir=${HOME}`
- `INPUT_KUBERNETES` - Enable kubernetes logs ingestion. Default: `true`.
- `INPUT_KUBERNETES_EXCLUDE_PATTERNS` - Comma separated list of log file path patterns to be excluded from processing. Example: `"*.gz", "*.tar"`. Default: `""`.
- `INPUT_JOURNALD` - Enable logs ingestion from journald. Default: `true`.
- `INPUT_KUBERNETES_AUDIT` - Enable kubernetes audit logs ingestion. Default: `true`.
- `OUTPUT_ELASTICSEARCH` - Enable logs output to Elasticsearch. Default `true`.
- `ELASTICSEARCH_HOST` - Elasticsearch host, can be comma separated. Default: `127.0.0.1:9200`.
- `ELASTICSEARCH_SCHEME`- Elasticsearch HTTP scheme. Default: `http`.
- `ELASTICSEARCH_SSL_ENABLED` - Elasticsearch SSL flag. Default: `false`.
- `ELASTICSEARCH_CA_CERTIFICATE_PATH` - The path to the .pem file that contains the Certificate Authority’s certificate.
- `ELASTICSEARCH_CERTIFICATE_VERIFICATION` - Elasticsearch SSL cerificate verification. Default: `true`.
- `ELASTICSEARCH_HTTP_COMPRESSION_ENABLED` - Elasticsearch HTTP compression. Default: `true`.
- `ELASTICSEARCH_USER` - Elasticsearch basic auth username. Default: `""`.
- `ELASTICSEARCH_PASSWORD` - Elasticsarch basic auth password. Default: `""`.
- `ELASTICSEARCH_INDEX_SUFFIX` - Elasticsearch index suffix. Default: `""`.
- `LOGSTASH_ARGS` - Sets additional logstash command line arguments.

For Kubernetes audit logs it may be necessary to increase `index.mapping.total_fields.limit`. This can be achieved with Elasticsearch curator [`index_settings` action](https://www.elastic.co/guide/en/elasticsearch/client/curator/5.1/index_settings.html).

## Running

```
$ docker run -ti --rm \
    -v /var/lib/logstash-kubernetes:/var/lib/logstash:z \
    -v /var/log/journal:/var/log/journal:ro \
    -v /var/lib/docker/containers:/var/lib/docker/containers:ro \
    -v /var/log/containers:/var/log/containers:ro \
    -v /var/log/kubernetes:/var/log/kubernetes:ro \
    -e ELASTICSEARCH_HOST=my-est-host.local:9200 \
    quay.io/ukhomeofficedigital/logstash-kubernetes:latest
```


CHANGELOG.md
# CHANGELOG

## 6.8.2-1

* Removes previously added `INPUT_KUBERNETES_FILE_CHUNK_COUNT`, `INPUT_KUBERNETES_FILE_CHUNK_SIZE` environment variables 
  due to the version of `logstash-input-file` plugin used.
* Preinstalls `logstash-input-file` plugin v4.0.5 due to log change tracking issues in v.4.1.x.
* Renames `logstash.yml` config option:
	`xpack.monitoring.elasticsearch.url` -> `xpack.monitoring.elasticsearch.hosts`

## 6.8.1-3

* Removes `ELASTICSEARCH_FLUSH_SIZE` as it's now obsolete.

## 6.8.1-2

* Adds the following environment variables:
	* `INPUT_KUBERNETES_FILE_CHUNK_COUNT` - [file_chunk_count](https://www.elastic.co/guide/en/logstash/6.8/plugins-inputs-file.html#plugins-inputs-file-file_chunk_count). Default: 32.
	* `INPUT_KUBERNETES_FILE_CHUNK_SIZE` - [file_chunk_size](https://www.elastic.co/guide/en/logstash/6.8/plugins-inputs-file.html#plugins-inputs-file-file_chunk_size). Default: 32768 (32KB).

	By default Logstash will read at most 1MB chunks at a time from each Docker container log file.

## 6.8.1-1

* Upgrades base image to fedora:30
* Upgrades Logstash to version 6.8.1: https://www.elastic.co/guide/en/logstash/6.8/releasenotes.html
* Bumps `journald` plugin version
* Installs `statsd` plugin as no longer part of the standard install.
* The following environment variables are ADDED:
	* `LS_NODE_NAME`
	* `LS_HTTP_HOST`
* Options added to README:
	* `LS_JAVA_OPTS`

## 5.6.16-1

* Upgrades Logstash to version 5.6.16: https://www.elastic.co/guide/en/logstash/5.6/releasenotes.html

* Added options:
	* `LS_MONITORING_ENABLE` - Whether to enable Logstash xpack monitoring. Default: `false`
	* `ELASTICSEARCH_SCHEME` - Elasticsearch HTTP scheme. Default: `http`.

### Notes:

`ELASTICSEARCH_SCHEME` should be set to `https` where appropriate, but is only actually used when LS_MONITORING_ENABLE is `true`.

Elasticsearch outputs with `document_type` are deprecated as they are not supported in version 6 or over: https://www.elastic.co/guide/en/elasticsearch/reference/6.0/removal-of-types.html

The following, for example, should be removed:

```
document_type => "kubernetes"
```

If in doubt, check Management > Upgrade Assistant > Indices in Kibana, which was added in v5.6, or the deprecation messages in log output.


run.sh
#!/usr/bin/bash

export HOME=/var/lib/logstash

: ${LS_LOG_LEVEL:=error}
: ${LS_HEAP_SIZE:=500m}
: ${LS_JAVA_OPTS:=-Djava.io.tmpdir=${HOME}}
: ${LS_LOG_DIR:=/var/lib/logstash}
: ${LS_OPEN_FILES:=8192}
: ${LS_PIPELINE_BATCH_SIZE:=125}
: ${LS_MONITORING_ENABLE:=false}
: ${LS_NODE_NAME:=$HOSTNAME}
: ${LS_HTTP_HOST:=0.0.0.0}
: ${INPUT_KUBERNETES_EXCLUDE_PATTERNS:=}
: ${INPUT_JOURNALD:=true}
: ${INPUT_KUBERNETES_AUDIT:=true}
: ${INPUT_KUBERNETES:=true}

: ${OUTPUT_ELASTICSEARCH:=true}
: ${ELASTICSEARCH_HOST:=127.0.0.1:9200}
: ${ELASTICSEARCH_SCHEME:=http}
: ${ELASTICSEARCH_SSL_ENABLED:=false}
: ${ELASTICSEARCH_CA_CERTIFICATE_PATH:=/etc/pki/tls/certs/ca-bundle.crt}
: ${ELASTICSEARCH_CERTIFICATE_VERIFICATION:=true}
: ${ELASTICSEARCH_USER:=}
: ${ELASTICSEARCH_PASSWORD:=}
: ${ELASTICSEARCH_HTTP_COMPRESSION_ENABLED:=true}

: ${ELASTICSEARCH_INDEX_SUFFIX:=""}

[ ${ELASTICSEARCH_SSL_ENABLED} == "true" ] && export ELASTICSEARCH_SCHEME="https"

# exclude certain kubernetes log files if provided
if [[ ${INPUT_KUBERNETES_EXCLUDE_PATTERNS} ]]; then
  sed -e "s/%INPUT_KUBERNETES_EXCLUDE_PATTERNS%/exclude => [ ${INPUT_KUBERNETES_EXCLUDE_PATTERNS} ]/" \
      -i /logstash/conf.d/10_input_kubernetes.conf
else
  sed -e "s/%INPUT_KUBERNETES_EXCLUDE_PATTERNS%//" \
      -i /logstash/conf.d/10_input_kubernetes.conf
fi

if [[ ${INPUT_JOURNALD} != 'true' ]]; then
  rm -f /logstash/conf.d/10_input_journald.conf
fi

if [[ ${INPUT_KUBERNETES_AUDIT} != 'true' ]]; then
  rm -f /logstash/conf.d/10_input_kubernetes_audit.conf
fi

if [[ ${INPUT_KUBERNETES} != 'true' ]]; then
  rm -f /logstash/conf.d/10_input_kubernetes.conf
fi

if [[ ${OUTPUT_ELASTICSEARCH} != 'true' ]]; then
  rm -f /logstash/conf.d/20_output_journald_elasticsearch.conf
  rm -f /logstash/conf.d/20_output_kubernetes_elasticsearch.conf
  rm -f /logstash/conf.d/20_output_kubernetes_audit_elasticsearch.conf
else
  sed -e "s/%ELASTICSEARCH_HOST%/${ELASTICSEARCH_HOST}/" \
      -e "s/%ELASTICSEARCH_SSL_ENABLED%/${ELASTICSEARCH_SSL_ENABLED}/" \
      -e "s#%ELASTICSEARCH_CA_CERTIFICATE_PATH%#${ELASTICSEARCH_CA_CERTIFICATE_PATH}#" \
      -e "s/%ELASTICSEARCH_CERTIFICATE_VERIFICATION%/${ELASTICSEARCH_CERTIFICATE_VERIFICATION}/" \
      -e "s/%ELASTICSEARCH_HTTP_COMPRESSION_ENABLED%/${ELASTICSEARCH_HTTP_COMPRESSION_ENABLED}/" \
      -e "s/%ELASTICSEARCH_USER%/${ELASTICSEARCH_USER}/" \
      -e "s/%ELASTICSEARCH_PASSWORD%/${ELASTICSEARCH_PASSWORD}/" \
      -e "s/%ELASTICSEARCH_INDEX_SUFFIX%/${ELASTICSEARCH_INDEX_SUFFIX}/" \
      -e "s/%ELASTICSEARCH_SCHEME%/${ELASTICSEARCH_SCHEME}/" \
      -e "s/%LS_MONITORING_ENABLE%/${LS_MONITORING_ENABLE}/" \
      -e "s/%LS_NODE_NAME%/${LS_NODE_NAME}/" \
      -e "s/%LS_HTTP_HOST%/${LS_HTTP_HOST}/" \
      -i /logstash/config/logstash.yml \
      -i /logstash/conf.d/20_output_kubernetes_elasticsearch.conf \
      -i /logstash/conf.d/20_output_kubernetes_audit_elasticsearch.conf \
      -i /logstash/conf.d/20_output_journald_elasticsearch.conf
fi


ulimit -n ${LS_OPEN_FILES} > /dev/null

exec /logstash/bin/logstash --log.format json \
  --log.level ${LS_LOG_LEVEL} \
  --pipeline.batch.size ${LS_PIPELINE_BATCH_SIZE} \
  --config.reload.automatic \
  -f "/logstash/conf.d/**/*.conf" \
  ${LOGSTASH_ARGS}
